Added option to exclude class methods or all parent methods

frankdekker opened the pull request:
Feature: Added ability to exclude specific methods or parent methods (or both)
Breaking: no

Added methods to ConstraintConfig:
```
::setExcludedMethods(array $methods)            // case sensitive excluded specific methods in the class
::setAssertParentMethods(bool)                  // if set to false, all parent methods will be excluded.
```

## Changes
- passed the `ConstraintConfig` to the constructor of `AccessorPairProvider` and `ConstructPairProvider`
- new class `ClassMethodProvider` to fetch (and excluded) methods of a class.
- added coverage for changed and new code.

frankdekker committed:
Added ability to exclude parent methods

frankdekker committed:
Add excluded methods to constraint config

frankdekker committed:
Add excluded methods to constraint config

frankdekker committed:
Add excluded methods to constructor method provider

frankdekker committed:
Add coverage for ClassMethodProvider

frankdekker committed:
Add coverage for ConstructorPairProvider

frankdekker committed:
Add coverage for ConstructorPairProvider

frankdekker committed:
Add coverage for ConstructorPairProvider

frankdekker committed:
Add coverage for ConstructorPairProvider

frankdekker committed:
Appease phpmd warning for AccessPairProvider

frankdekker committed:
Appease phpmd warning for AccessPairProvider

frankdekker commented on src/Constraint/MethodPair/AccessorPair/AccessorPairProvider.php:
```diff
@@ -85,6 +79,28 @@ protected function getMethodBaseNames(string $methodName, string $getterPrefix):
+    /**
+     * @return ReflectionMethod[]
+     */
+    protected function getSetters(ReflectionClass $class, string $baseMethodName): array
```
I hit the phpmd cyclomatic complexity. Extracted this code to separate method

frankdekker committed:
Fix docblock comment

frankdekker committed:
Add extra trait utility asserts functions

frankdekker committed:
Add extra trait utility asserts functions

bram123 commented:
@frankdekker 
can you add these extra tests:
- parent has getter, child has setter
- parent has setter, child has getter
- parent has constructor, child has getter
- parent has getter, child has constructor
With allowParentMethods true & false

And probably best to group these child/parent tests into a data/parentchild folder.

frankdekker committed:
Add extra parent child setter,getter tests

frankdekker committed:
Add extra parent child setter,getter tests

frankdekker committed:
Reorganize data directory

frankdekker committed:
Reorganize data directory

frankdekker committed:
Added extra coverage for parent constructors/method scenarios

frankdekker committed:
Added extra coverage for parent constructors/method scenarios

frankdekker committed:
Rename data to fixtures

frankdekker commented:
> 
> 
> @frankdekker
> can you add these extra tests:
> 
>     * parent has getter, child has setter
> 
>     * parent has setter, child has getter
> 
>     * parent has constructor, child has getter
> 
>     * parent has getter, child has constructor
>       With allowParentMethods true & false
> 
> 
> And probably best to group these child/parent tests into a data/parentchild folder.

Added subfolder with parentchild.
Added tests for the scenario's above.

frankdekker committed:
Revert "Rename data to fixtures"

This reverts commit bab22ead
